File: ChatBox/1. PresentationLayer/Modules/ThemesModule/Controller/objC (switched off)/ThemesViewController.m
//
//  ThemesViewController.m
//  ChatBox
//
//  Created by Alexey Teplonogov on 11/10/2018.
//  Copyright © 2018 Alexey Teplonogov. All rights reserved.
//

#import "ThemesViewController.h"
#import "UIColor+ChatBox.h"

@interface ThemesViewController () 

@property (retain, nonatomic) Themes* themes;

@end

@implementation ThemesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    NSData* themeData = [NSUserDefaults.standardUserDefaults objectForKey:@"theme"];
    
    if (themeData != nil) {
        UIColor* themeColor = [NSKeyedUnarchiver unarchiveObjectWithData:themeData];
        self.view.backgroundColor = themeColor;
    } else {
        self.navigationController.navigationBar.tintColor = [UIColor colorWithRed:96.0f/255.0f green:154/255.0f blue:234/255.0f alpha:1];
        self.view.backgroundColor = UIColor.whiteColor;
    }
    
    UIImage* closeButtonImage = [UIImage imageNamed:@"CloseButton"];
    
    UIBarButtonItem* closeButton = [[[UIBarButtonItem alloc] initWithImage:closeButtonImage style:UIBarButtonItemStyleDone target:self action:@selector(closeAction)] autorelease];
    
    _themeOneButton.layer.cornerRadius = 10;
    _themeTwoButton.layer.cornerRadius = 10;
    _themeThreeButton.layer.cornerRadius = 10;
    
    self.themes = [[[Themes alloc] init] autorelease];
    self.themes.theme1 = [UIColor whiteColor];
    self.themes.theme2 = [UIColor skyBlueColor];
    self.themes.theme3 = [UIColor nightBlueColor];
        
    self.navigationController.navigationBar.topItem.leftBarButtonItem = closeButton;
    
    self.navigationController.navigationBar.translucent = NO;
    
}

- (void)dealloc {
    [self.themes release];
    //self.themes = nil;


    NSLog(@"controller deallocated");
    [super dealloc];
}

#pragma mark - Actions

- (void)closeAction {
    [self dismissViewControllerAnimated:true completion:nil];
}

- (IBAction)themeOneAction:(UIButton *)sender {
    [UIView animateWithDuration:0.3 animations:^{
        self.view.backgroundColor = self.themes.theme1;
        self.navigationController.navigationBar.tintColor = UIColor.skyBlueColor;
        [self.navigationController.navigationBar setBarStyle:UIBarStyleDefault];
        [self setNavBarTheme:self.themes.theme1];
    }];
    [self.delegate themesViewController:self didSelectTheme:self.themes.theme1];
}

- (IBAction)themeTwoAction:(UIButton *)sender {
    [UIView animateWithDuration:0.3 animations:^{
        self.view.backgroundColor = self.themes.theme2;
        self.navigationController.navigationBar.tintColor = UIColor.whiteColor;
        [self.navigationController.navigationBar setBarStyle:UIBarStyleBlackOpaque];
        [self setNavBarTheme:self.themes.theme2];
    }];
    
    [self.delegate themesViewController:self didSelectTheme:self.themes.theme2];
}

- (IBAction)themeThreeAction:(id)sender {
    [UIView animateWithDuration:0.3 animations:^{
        self.view.backgroundColor = self.themes.theme3;
        self.navigationController.navigationBar.tintColor = UIColor.whiteColor;
        [self.navigationController.navigationBar setBarStyle:UIBarStyleBlackOpaque];
        [self setNavBarTheme:self.themes.theme3];
    }];
    
    [self.delegate themesViewController:self didSelectTheme:self.themes.theme3];
}

#pragma mark - Helpers

- (void)setNavBarTheme:(UIColor*) color {
    self.navigationController.navigationBar.barTintColor = color;
}


@end

File: ChatBox/1. PresentationLayer/Modules/ThemesModule/Controller/ThemesViewController.swift
//
//  ThemesViewController.swift
//  ChatBox
//
//  Created by Alexey Teplonogov on 14/10/2018.
//  Copyright © 2018 Alexey Teplonogov. All rights reserved.
//

import UIKit

class ThemesViewController: UIViewController {

    @IBOutlet var themeOneButton: UIButton!
    @IBOutlet var themeTwoButton: UIButton!
    @IBOutlet var themeThreeButton: UIButton!

    // Dependencies

    private let model: IThemesModel
    private let presentationAssembly: IPresentationAssembly

    init(model: IThemesModel, presentationAssembly: PresentationAssembly) {
        self.model = model
        self.presentationAssembly = presentationAssembly
        super.init(nibName: "ThemesViewController", bundle: nil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    var handler: ((UIColor) -> Void)!

    override func viewDidLoad() {
        super.viewDidLoad()

        model.setTheme(forVC: self)

        let closeButtonImage = #imageLiteral(resourceName: "CloseButton")

        let closeButton = UIBarButtonItem.init(image: closeButtonImage,
                                               style: .done,
                                               target: self,
                                               action: #selector(closeAction))

        themeOneButton.layer.cornerRadius = 10
        themeTwoButton.layer.cornerRadius = 10
        themeThreeButton.layer.cornerRadius = 10

        model.themes.theme1 = UIColor.white
        model.themes.theme2 = UIColor.skyBlue()
        model.themes.theme3 = UIColor.nightBlue()

        self.navigationController?.navigationBar.topItem?.leftBarButtonItem = closeButton
        self.navigationController?.navigationBar.isTranslucent = false
        self.title = "Themes"
    }

    // MARK: - Actions

    @objc func closeAction() {
        self.dismiss(animated: true, completion: nil)
    }

    @IBAction func themeOneAction(_ sender: UIButton) {

        UIView.animate(withDuration: 0.3) {
            self.view.backgroundColor = self.model.themes.theme1
            self.navigationController?.navigationBar.tintColor = UIColor.skyBlue()
            self.navigationController?.navigationBar.barStyle = .default
            self.setNavBarTheme(color: self.model.themes.theme1)
        }

        handler(model.themes.theme1)

    }

    @IBAction func themeTwoAction(_ sender: UIButton) {
        UIView.animate(withDuration: 0.3) {
            self.view.backgroundColor = self.model.themes.theme2
            self.navigationController?.navigationBar.tintColor = UIColor.white
            self.navigationController?.navigationBar.barStyle = .blackOpaque
            self.setNavBarTheme(color: self.model.themes.theme2)
        }

        handler(model.themes.theme2)
    }

    @IBAction func themeThreeAction(_ sender: UIButton) {
        UIView.animate(withDuration: 0.3) {
            self.view.backgroundColor = self.model.themes.theme3
            self.navigationController?.navigationBar.tintColor = UIColor.white
            self.navigationController?.navigationBar.barStyle = .blackOpaque
            self.setNavBarTheme(color: self.model.themes.theme3)
        }

        handler(model.themes.theme3)
    }

    func setNavBarTheme(color: UIColor) {
        self.navigationController?.navigationBar.barTintColor = color
    }

}
